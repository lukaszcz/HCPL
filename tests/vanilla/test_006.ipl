
let f = \!x {
   let g = \y . + x y;
   if = x 0 then (f 10) (g 5)
   else if < x 10 then g (f (- x 1))
   else g
};

let g = \!x \!h {
   + x (h x)
};

let h = \!x if = x 0 then 0 else g (- x 1) h;

print ((f 100) 500); // 600
print (f 5); // 30
print (g 10 h); // 55
print (g 3 (f 10)); // 16

;;;

let f2 = \&x {
   let f = \&y . + x y;
   if = x 0 then (f2 10) (f 5)
   else if < x 10 then f (f2 (- x 1))
   else f
};

let g2 = \&x \&h {
   + x (h x)
};

let h2 = \&x if = x 0 then 0 else g2 (- x 1) h2;

print ((f2 100) 500); // 600
print (f2 5); // 30
print (g2 10 h2); // 55
print (g2 3 (f2 10)); // 16

;;;
