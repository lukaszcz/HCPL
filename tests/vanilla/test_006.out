600
30
55
16
600
30
55
16
